File: internal/estimateaudit/parity.go
// Per-joinrel parity against the PG 18.3 reference — the bar
// docs/design/leftdeep-joins/09-verification-and-acceptance.md §4 asks P5.9 to
// certify, restated after the 2026-08-05 reference measurement.
//
// Why the absolute tripwire of §5 cannot be that bar: on TPC-H Q18 **PG itself
// misestimates the final joinrel by 1 674×** (it dedups the GROUP BY subquery
// to a plain inner join and estimates 117 159 against an actual of 70), which
// is over §5's own 10³ tripwire. A milestone whose goal is PG-shaped planning
// cannot be gated on a bar the reference implementation fails: the absolute
// factor answers "is this estimate good?", and the question P5.9 has to answer
// is "is this estimate WORSE THAN PG's?". Those differ wherever the query
// itself is hard to estimate — exactly where a ratchet is worth having.
//
// The unit of comparison is the JOINREL, identified the way upstream
// identifies it: by the set of base relations underneath it
// (`RelOptInfo.relids`). Two plans with different join ORDERS still contain the
// same joinrel whenever they build the same relation set, and the ACTUAL row
// count of that joinrel is a property of the query and the data, not of the
// plan — so goopg's misestimate factor for {customer,orders} and PG's for
// {customer,orders} are the same estimation problem even when the two engines
// reached it by different spines. A joinrel only one engine built is NOT a
// parity failure; it is a shape divergence, counted separately and classed per
// §6.
package estimateaudit

import (
	"fmt"
	"math"
	"sort"
	"strings"
)

// DefaultParitySlack is how many times worse than the reference goopg's
// misestimate may be before the ratchet counts it. It is not 1 because §4
// deliberately declines a match-all bar "while cost constants and stats
// fidelity still differ" — the ratchet exists to make DRIFT visible, and an
// order of magnitude past the reference is drift no stats-fidelity gap
// explains.
const DefaultParitySlack = 10.0

// DefaultParityFloor keeps the ratchet off joinrels that are accurate in
// absolute terms. Without it a joinrel PG estimates exactly (ratio 1.0) and
// goopg estimates within 20× would count as a 20× parity failure, which says
// nothing about join order at that size.
const DefaultParityFloor = 100.0

// ParityBar parameterises §4's ratchet.
type ParityBar struct {
	Slack float64 // goopg ratio / reference ratio must stay under this
	Floor float64 // ... and goopg's own ratio must exceed this to count
}

// DefaultParityBar is §4's ratchet as restated by P5.6-g-iii.
func DefaultParityBar() ParityBar {
	return ParityBar{Slack: DefaultParitySlack, Floor: DefaultParityFloor}
}

// ParityStatus classifies one joinrel of one query.
type ParityStatus string

const (
	ParityMatched   ParityStatus = "matched"    // both engines built this joinrel
	ParityGoopgOnly ParityStatus = "goopg-only" // shape divergence (§6 class (b))
	ParityRefOnly   ParityStatus = "pg-only"    // shape divergence (§6 class (b))
)

// ParityRow is one joinrel, on one or both sides.
type ParityRow struct {
	Query  string
	Rels   []string
	Status ParityStatus
	Goopg  *Node // nil when Status is ParityRefOnly
	Ref    *Node // nil when Status is ParityGoopgOnly
	Final  bool  // this is goopg's final joinrel (or the reference's, for pg-only rows)

	// Ambiguous marks a key that more than one join level in the same plan
	// resolved to — see joinsByRelKey. The row still compares the outermost
	// occurrence, but the reader is told, because the printed plan does not
	// contain enough information to separate the two.
	Ambiguous bool
}

func (r ParityRow) goopgRatio() float64 {
	if r.Goopg == nil {
		return 0
	}
	return r.Goopg.Ratio()
}

func (r ParityRow) refRatio() float64 {
	if r.Ref == nil {
		return 0
	}
	return r.Ref.Ratio()
}

// Excess is the ratio-of-ratios: how many times worse than the reference
// goopg's estimate of this joinrel is. 1.0 means "as good as PG"; below 1.0
// means goopg is the more accurate of the two. Unmatched rows have no excess.
func (r ParityRow) Excess() float64 {
	if r.Status != ParityMatched {
		return 0
	}
	ref := math.Max(r.refRatio(), 1)
	return r.goopgRatio() / ref
}

// Violates reports whether this row trips the ratchet: goopg materially worse
// than the reference (Slack) AND materially wrong in absolute terms (Floor).
// Both conditions, because either alone flags something the ratchet does not
// mean — see DefaultParityFloor.
func (r ParityRow) Violates(p ParityBar) bool {
	if r.Status != ParityMatched {
		return false
	}
	return r.Excess() > p.Slack && r.goopgRatio() > p.Floor
}

func (r ParityRow) String() string {
	rels := strings.Join(r.Rels, "+")
	switch r.Status {
	case ParityGoopgOnly:
		return fmt.Sprintf("%s: {%s} built by goopg only (est=%d actual=%.0f ratio=%.1fx); PG built no such joinrel",
			r.Query, rels, r.Goopg.EstRows, r.Goopg.ActualTotal(), r.goopgRatio())
	case ParityRefOnly:
		return fmt.Sprintf("%s: {%s} built by PG only (est=%d actual=%.0f ratio=%.1fx); goopg built no such joinrel",
			r.Query, rels, r.Ref.EstRows, r.Ref.ActualTotal(), r.refRatio())
	default:
		return fmt.Sprintf("%s: {%s} goopg est=%d actual=%.0f ratio=%.1fx vs PG est=%d actual=%.0f ratio=%.1fx — %.1fx worse than the reference",
			r.Query, rels, r.Goopg.EstRows, r.Goopg.ActualTotal(), r.goopgRatio(),
			r.Ref.EstRows, r.Ref.ActualTotal(), r.refRatio(), r.Excess())
	}
}

// Parity joins the two engines' audits per query and per joinrel. Queries
// unmeasured on either side are skipped (a missing reference is not a pass and
// not a failure — RenderParity names them).
func Parity(goopg, reference []QueryReport) []ParityRow {
	refByName := make(map[string]QueryReport, len(reference))
	for _, r := range reference {
		refByName[r.Name] = r
	}
	var out []ParityRow
	for _, g := range goopg {
		ref, ok := refByName[g.Name]
		if !ok || g.Err != "" || ref.Err != "" {
			continue
		}
		gJoins, gDup := joinsByRelKey(g)
		rJoins, rDup := joinsByRelKey(ref)
		for _, key := range sortedKeys(gJoins) {
			gn := gJoins[key]
			row := ParityRow{Query: g.Name, Rels: gn.Rels, Goopg: gn, Status: ParityGoopgOnly,
				Final:     g.Final != nil && g.Final.RelKey() == key,
				Ambiguous: gDup[key] > 1}
			if rn, ok := rJoins[key]; ok {
				row.Ref = rn
				row.Status = ParityMatched
				row.Ambiguous = row.Ambiguous || rDup[key] > 1
			}
			out = append(out, row)
		}
		for _, key := range sortedKeys(rJoins) {
			if _, ok := gJoins[key]; ok {
				continue
			}
			rn := rJoins[key]
			out = append(out, ParityRow{Query: g.Name, Rels: rn.Rels, Ref: rn, Status: ParityRefOnly,
				Final:     ref.Final != nil && ref.Final.RelKey() == key,
				Ambiguous: rDup[key] > 1})
		}
	}
	return out
}

// joinsByRelKey indexes a report's joins by joinrel identity, and counts how
// many join levels resolved to each key. A join with no ANALYZE
// instrumentation is dropped: it has no actual to compare.
//
// A repeated key is not a bug in the caller but a limit of the printed plan.
// TPC-H Q18 is the live case: the subquery's `lineitem` is a different range
// table entry from the outer query's, but goopg prints both as
// `Seq Scan on public.lineitem` with no alias, so the SEMI above them keys the
// same as the inner join below it. Upstream's relids would separate them; the
// text cannot. The outermost occurrence wins (final-joinrel comparisons are
// what §4 is stated on) and the collision is surfaced via ParityRow.Ambiguous
// rather than silently resolved.
func joinsByRelKey(r QueryReport) (map[string]*Node, map[string]int) {
	out := make(map[string]*Node, len(r.Joins))
	dup := make(map[string]int, len(r.Joins))
	for i := range r.Joins {
		n := &r.Joins[i]
		if !n.HasAct || len(n.Rels) == 0 {
			continue
		}
		key := n.RelKey()
		dup[key]++
		if prev, ok := out[key]; ok && prev.Depth <= n.Depth {
			continue
		}
		out[key] = n
	}
	return out, dup
}

// ParityViolations lists the ratchet failures, worst first.
func ParityViolations(rows []ParityRow, p ParityBar) []ParityRow {
	var out []ParityRow
	for _, r := range rows {
		if r.Violates(p) {
			out = append(out, r)
		}
	}
	sort.SliceStable(out, func(i, j int) bool { return out[i].Excess() > out[j].Excess() })
	return out
}

// ParityMismatches lists the shape divergences (joinrels only one engine
// built). §4 counts these into the pinned mismatch budget; they are not
// parity violations because there is nothing to compare.
func ParityMismatches(rows []ParityRow) []ParityRow {
	var out []ParityRow
	for _, r := range rows {
		if r.Status != ParityMatched {
			out = append(out, r)
		}
	}
	return out
}

// RenderParity writes the parity column of the committed artifact: one line
// per joinrel with both engines' ratios and the excess, then the ratchet
// summary. Deterministic, like Render.
func RenderParity(goopg, reference []QueryReport, rows []ParityRow, p ParityBar) string {
	var b strings.Builder
	b.WriteString("=== PER-JOINREL PARITY vs PG 18.3 (09 §4 ratchet)\n")

	byQuery := map[string][]ParityRow{}
	for _, r := range rows {
		byQuery[r.Query] = append(byQuery[r.Query], r)
	}
	for _, q := range sortedKeys(byQuery) {
		fmt.Fprintf(&b, "--- %s\n", q)
		for _, r := range byQuery[q] {
			mark := "  "
			if r.Final {
				mark = "F "
			}
			if r.Ambiguous {
				mark = mark[:1] + "~" // "F~" / " ~", still two columns wide
			}
			switch r.Status {
			case ParityMatched:
				fmt.Fprintf(&b, "  %s%-46s goopg %8.1fx | PG %8.1fx | excess %7.1fx%s\n",
					mark, truncate("{"+strings.Join(r.Rels, "+")+"}", 46),
					r.goopgRatio(), r.refRatio(), r.Excess(), violMark(r, p))
			case ParityGoopgOnly:
				fmt.Fprintf(&b, "  %s%-46s goopg %8.1fx | PG        - | SHAPE (goopg-only joinrel)\n",
					mark, truncate("{"+strings.Join(r.Rels, "+")+"}", 46), r.goopgRatio())
			case ParityRefOnly:
				fmt.Fprintf(&b, "  %s%-46s goopg        - | PG %8.1fx | SHAPE (PG-only joinrel)\n",
					mark, truncate("{"+strings.Join(r.Rels, "+")+"}", 46), r.refRatio())
			}
		}
	}

	v := ParityViolations(rows, p)
	m := ParityMismatches(rows)
	var ambiguous int
	for _, r := range rows {
		if r.Ambiguous {
			ambiguous++
		}
	}
	fmt.Fprintf(&b, "\n=== PARITY SUMMARY (slack %.0fx, floor %.0fx)\n", p.Slack, p.Floor)
	fmt.Fprintf(&b, "  joinrels compared: %d matched, %d shape-divergent, %d ambiguous "+
		"(a relation scanned twice without an alias — outermost occurrence compared)\n",
		len(rows)-len(m), len(m), ambiguous)
	if len(v) == 0 {
		b.WriteString("  no joinrel worse than the reference beyond the ratchet\n")
	}
	for _, r := range v {
		fmt.Fprintf(&b, "  PARITY-VIOLATION %s\n", r)
	}
	fmt.Fprintf(&b, "  RATCHET parity_violations=%d shape_mismatches=%d\n", len(v), len(m))
	for _, name := range uncomparedQueries(goopg, reference) {
		fmt.Fprintf(&b, "  UNCOMPARED %s\n", name)
	}
	return b.String()
}

func violMark(r ParityRow, p ParityBar) string {
	if r.Violates(p) {
		return "  <== PARITY-VIOLATION"
	}
	return ""
}

// uncomparedQueries names every query the parity gate could not compare, with
// the side that is missing. §5's rule applies here too: a query silently
// absent from the comparison reads as a compared-and-clean one.
func uncomparedQueries(goopg, reference []QueryReport) []string {
	refByName := map[string]QueryReport{}
	for _, r := range reference {
		refByName[r.Name] = r
	}
	var out []string
	for _, g := range goopg {
		ref, ok := refByName[g.Name]
		switch {
		case !ok:
			out = append(out, g.Name+": no reference plan captured")
		case g.Err != "":
			out = append(out, g.Name+": goopg unmeasured ("+g.Err+")")
		case ref.Err != "":
			out = append(out, g.Name+": reference unmeasured ("+ref.Err+")")
		}
	}
	return out
}

// attachRels computes every node's joinrel identity from the printed plan.
//
// The tree is reconstructed from indentation alone, which is why the depth
// STEP must never be assumed: goopg indents two spaces per level and upstream
// six, so the rule is "strictly deeper than mine, before the next node at my
// depth or shallower, is my descendant" rather than any arithmetic on depth.
func attachRels(nodes []Node) {
	if len(nodes) == 0 {
		return
	}
	parent := make([]int, len(nodes))
	var stack []int
	for i := range nodes {
		for len(stack) > 0 && nodes[stack[len(stack)-1]].Depth >= nodes[i].Depth {
			stack = stack[:len(stack)-1]
		}
		parent[i] = -1
		if len(stack) > 0 {
			parent[i] = stack[len(stack)-1]
		}
		stack = append(stack, i)
	}
	// Children always follow their parent in the printed order, so one
	// backwards pass propagates every leaf to every ancestor.
	for i := len(nodes) - 1; i >= 0; i-- {
		if rel, ok := leafRel(nodes[i].Label); ok {
			nodes[i].Rels = append(nodes[i].Rels, rel)
		}
		nodes[i].Rels = dedupeSorted(nodes[i].Rels)
		if p := parent[i]; p >= 0 {
			nodes[p].Rels = append(nodes[p].Rels, nodes[i].Rels...)
		}
	}
}

// leafRel extracts the base relation a scan node reads, in a form comparable
// across the two engines' renderings:
//
//	goopg: "Seq Scan on public.lineitem l1 (stats)" / "Index Scan using public.orders_pk on public.orders"
//	PG:    "Seq Scan on lineitem l1"                / "Index Scan using orders_pkey on orders"
//
// The ALIAS wins when present, because that is what distinguishes the three
// lineitem scans of Q21 — without it a self-join's joinrels collapse into one
// key. Schema qualification and goopg's "(stats)" marker are dropped: they are
// rendering differences, not identity.
func leafRel(label string) (string, bool) {
	// A bitmap index scan names an INDEX, not a relation; its parent bitmap
	// heap scan supplies the relation. Counting it would invent a leaf.
	if strings.HasPrefix(label, "Bitmap Index Scan") {
		return "", false
	}
	i := strings.LastIndex(label, " on ")
	if i < 0 {
		return "", false
	}
	rest := strings.TrimSpace(label[i+len(" on "):])
	rest = strings.ReplaceAll(rest, "(stats)", "")
	fields := strings.Fields(rest)
	if len(fields) == 0 {
		return "", false
	}
	name := fields[0]
	if len(fields) > 1 && !strings.HasPrefix(fields[1], "(") {
		name = fields[1] // alias
	}
	if j := strings.LastIndex(name, "."); j >= 0 {
		name = name[j+1:]
	}
	name = strings.Trim(name, `"`)
	if name == "" {
		return "", false
	}
	return name, true
}

func dedupeSorted(in []string) []string {
	if len(in) < 2 {
		return in
	}
	sort.Strings(in)
	out := in[:1]
	for _, s := range in[1:] {
		if s != out[len(out)-1] {
			out = append(out, s)
		}
	}
	return out
}

// SplitPlansFile splits a `<label>.plans.txt` artifact (written by
// cmd/estimate-audit --keep-plans) back into per-query plan text, so a parity
// run can be re-derived offline from committed evidence instead of re-running
// a 13-minute power run.
func SplitPlansFile(text string) []QueryReport {
	var out []QueryReport
	var name string
	var body strings.Builder
	flush := func() {
		if name == "" {
			return
		}
		out = append(out, Audit(name, body.String()))
		body.Reset()
	}
	for _, line := range strings.Split(text, "\n") {
		if strings.HasPrefix(line, "=== ") {
			flush()
			name = strings.TrimSpace(strings.TrimPrefix(line, "=== "))
			continue
		}
		body.WriteString(line)
		body.WriteString("\n")
	}
	flush()
	return out
}

File: scripts/estimate-parity/parity.py
#!/usr/bin/env python3
"""The estimate-accuracy (EA) parity ratchet, over TPC-DS.

  usage: parity.py <goopg-EXPLAIN-ANALYZE-capture> <pg-plan-dir> [--json OUT]
                   [--baseline B] [--write-baseline B] [--tol T] [--floor F]

WHY THIS EXISTS. Four TODO_ALL items (C-05, C-10a, C-20a, C-21) cite an
"EA ratchet" as their acceptance gate. Ledger `take3-ea-ratchet-never-ran`
established that it has never run: `scripts/tpch-estimate-audit-arm.sh` is
invoked by no Makefile target, hook, precommit script or nightly stage, and
its default pinned PG baseline is absent from the tree, so a default-flag run
exits before it measures anything. Three structural gaps made it unusable for
the defects it was cited against even if it had been wired: it is TPC-H only
(cmd/estimate-audit/main.go:317-322), it is joinrel-granular so a base
relation estimating `rows=1` is not a candidate (audit.go:96-98), and TPC-H
carries no LIMIT, so the NLI+Memoize shape that produces the aggregate
over-estimates never arises.

This tool closes all four gaps. It reads a real `EXPLAIN (ANALYZE)` capture
over the TPC-DS SF0.5 corpus, so its truth is measured rather than estimated;
it scores base relations and joinrels alike; and its bar is PG-RELATIVE.

WHY THE BAR MUST BE PG-RELATIVE, AND NOT ABSOLUTE. An absolute est-vs-actual
bar is wrong in both directions on this corpus. Q47/Q57/Q81/Q89 emit
`rows=1` on a node whose actual is tens of thousands, and PostgreSQL 18.3
emits `rows=1` on the very same node (bench/tpcds/plans-pg/Q47.txt): a chain
of mutually-implied equi-join selectivities multiplied independently, which
upstream does not de-duplicate at estimate time either. Those are not
defects, and an absolute bar fails them — which is how a gate gets disabled.
Conversely Q99's 8007x aggregate over-estimate is a real defect PG does not
share, and a loose absolute bar passes it. The only bar that PASSES Q47 and
FAILS Q99 is "goopg is materially worse than PostgreSQL on this node".

THE KEY. Nodes are matched across the two planners by the SET OF BASE
RELATIONS in their subtree, which is the one coordinate both planners agree
on: it survives a different join order, a different join algorithm, an extra
Gather, and a materialisation the other side does not have. A singleton set
is a base relation; a larger set is a joinrel; so one keying gives both
granularities, which is exactly what the joinrel-only predecessor could not
do. CTE bodies and sub-plans are keyed under a pseudo-relation named for the
CTE, so a CTE reference does not silently unify with the query that fills it.

THE SCORE. Symmetric q-error, `max(est, actual) / min(est, actual)` with both
floored at 1 — the standard cardinality-estimation metric, which treats a
1000x under-estimate and a 1000x over-estimate as equally bad (they are: one
picks a nested loop over a hash join, the other the reverse).

Estimates and actuals are both taken PER LOOP, never multiplied out: `rows=`
under the inner side of a nested loop is what the planner expects per
rescan, and PG's own annotation means the same thing. Multiplying the actual
by `loops` and comparing it against a per-loop estimate manufactures an
error equal to the loop count on every parameterised path.
"""

import argparse
import json
import os
import re
import sys

# A plan line: name, then PG's cost annotation, then optionally the ANALYZE
# annotation. Written to accept `never executed` because a node under an
# unexecuted SubPlan legitimately has no actual and must be dropped, not
# scored as zero.
NODE = re.compile(
    r'^(?P<pre>\s*)(?P<arrow>->\s{2})?(?P<name>[A-Z].*?)\s+'
    r'\(cost=(?P<sc>[\d.]+)\.\.(?P<tc>[\d.]+) rows=(?P<erows>\d+) width=(?P<w>\d+)\)'
    # `actual time=` is present only under TIMING ON; the capture runs with
    # TIMING OFF (the gate scores row counts, not milliseconds, and TIMING ON
    # costs a clock_gettime per tuple on a 99-query sweep), so the time pair
    # must be OPTIONAL. Requiring it matched zero nodes in the whole corpus
    # and reported a clean gate over an empty population.
    r'(?:\s+\((?:(?:actual\s+(?:time=(?P<st>[\d.]+)\.\.(?P<et>[\d.]+)\s+)?'
    r'rows=(?P<arows>[\d.]+)\s+loops=(?P<loops>[\d.]+))|never executed)\))?\s*$')

# `CTE v1` / `InitPlan 1 (returns $0)` / `SubPlan 2` marker lines. These carry
# no cost annotation but DO open an indent level, so a parser that ignores
# them attaches a CTE body to whatever node happened to precede it and every
# relation-set above that point is wrong.
MARK = re.compile(r'^(?P<pre>\s*)(?P<kind>CTE|InitPlan|SubPlan)\s+(?P<label>\S+).*$')

# Base-relation extraction. The trailing alias is dropped deliberately: goopg
# and PG do not always agree on when an alias is printed, and the alias is not
# part of the coordinate the key is built on.
SCAN_ON = re.compile(r'\bon\s+(?:(?P<schema>\w+)\.)?(?P<rel>\w+)')
CTE_SCAN = re.compile(r'^CTE Scan on (?P<rel>\w+)')
WORKTABLE = re.compile(r'^WorkTable Scan on (?P<rel>\w+)')

# Node names that scan something nameable. "Subquery Scan on x" is excluded on
# purpose: `x` is an inlined subquery alias, not a relation, and its subtree
# already contributes the real relations underneath it.
SCAN_KINDS = ('Seq Scan', 'Index Scan', 'Index Only Scan', 'Bitmap Heap Scan',
              'Tid Scan', 'Sample Scan', 'Foreign Scan')

# `Bitmap Index Scan on customer_pkey` names an INDEX, not a relation, and it
# contains the substring `Index Scan`. Without this exclusion the index name
# joins the relation set and a bitmap plan's joinrel keys stop matching the
# same joinrel reached by any other access method — on either side.


class Node:
    __slots__ = ('name', 'erows', 'width', 'arows', 'loops', 'indent',
                 'parent', 'children', 'relset', 'scope', 'pseudo')

    def __init__(self, name, erows, width, arows, loops, indent, scope,
                 pseudo=False):
        self.name = name
        self.erows = erows
        self.width = width
        self.arows = arows
        self.loops = loops
        self.indent = indent
        self.scope = scope
        self.parent = None
        self.children = []
        self.relset = None
        # A `CTE v1` / `SubPlan 2` marker carries no cost annotation of its
        # own; it exists to open a scope. It must never be SCORED, or every
        # CTE contributes a phantom node with est=0.
        self.pseudo = pseudo


def base_relation(name, scope):
    """The base relation this node scans, qualified by its CTE/SubPlan scope.

    Returns None for a node that scans nothing nameable. The scope prefix is
    what stops a CTE body's `store_sales` from unifying with the main query's
    `store_sales`: they are different instances with different quals, and
    merging them would compare a filtered estimate against an unfiltered
    actual.
    """
    m = CTE_SCAN.match(name) or WORKTABLE.match(name)
    if m:
        # A CTE reference is a leaf here, not a window onto the body: the body
        # is scored separately under its own scope. Naming it `cte:v1` keeps
        # goopg's inlined clone and PG's shared body on the same key.
        return 'cte:' + m.group('rel')
    head = name.split('  ')[0].strip()
    if 'Bitmap Index Scan' in head:
        return None
    for kind in SCAN_KINDS:
        if kind in head:
            m = SCAN_ON.search(head)
            if m:
                return (scope + '.' if scope else '') + m.group('rel')
            return None
    return None


def parse(body):
    """Rebuild the plan forest from raw indentation.

    Marker lines push a pseudo-level and rename the scope, so everything under
    `CTE v1` is keyed `v1.<rel>`.
    """
    roots, stack = [], []
    for ln in body.splitlines():
        m = NODE.match(ln)
        if m:
            indent = len(m.group('pre'))
            # A non-`->` line indented past column 2 is a detail line
            # (`Sort Key:`, `Filter:`) that happens to start with a capital.
            if not m.group('arrow') and indent > 2:
                continue
            while stack and stack[-1][0] >= indent:
                stack.pop()
            parent = stack[-1][1] if stack else None
            n = Node(m.group('name').strip(), int(m.group('erows')),
                     int(m.group('w')),
                     float(m.group('arows')) if m.group('arows') else None,
                     float(m.group('loops')) if m.group('loops') else None,
                     indent, parent.scope if parent else '')
            if parent is not None:
                n.parent = parent
                parent.children.append(n)
            else:
                roots.append(n)
            stack.append((indent, n))
            continue
        mk = MARK.match(ln)
        if mk:
            indent = len(mk.group('pre'))
            while stack and stack[-1][0] >= indent:
                stack.pop()
            kind, label = mk.group('kind'), mk.group('label')
            scope = label if kind == 'CTE' else '%s%s' % (kind.lower(), label)
            # A marker is a PSEUDO-NODE and a new ROOT, not a child of the
            # node above it. Attaching a CTE body under the plan node that
            # happens to precede it would fold the body's relations into
            # every ancestor's relation set, so a `CTE v1` scanning
            # store_sales would make the plan root's key contain both the
            # main query's store_sales and the body's — two different
            # instances, unified, and every joinrel key above wrong.
            p = Node(kind + ' ' + label, 0, 0, None, None, indent, scope,
                     pseudo=True)
            roots.append(p)
            stack.append((indent, p))
    return roots


def walk(n):
    yield n
    for c in n.children:
        yield from walk(c)


def annotate_relsets(n):
    """Bottom-up: every node's key is the set of base relations beneath it."""
    s = set()
    for c in n.children:
        s |= annotate_relsets(c)
    own = base_relation(n.name, n.scope)
    if own:
        s.add(own)
    n.relset = frozenset(s)
    return s


def qerror(est, actual):
    """Symmetric q-error with both sides floored at 1.

    The floor is not cosmetic: PG clamps every estimate to >= 1
    (`clamp_row_est`), and goopg's `EstimateRows` does the same, so a node
    that truly returns 0 rows would otherwise divide by zero on the side the
    planner cannot express anyway.
    """
    e = max(float(est), 1.0)
    a = max(float(actual), 1.0)
    return max(e / a, a / e)


def collect(roots, want_actual):
    """{relset: {'est':…, 'actual':…, 'node':…}} for one query's plan forest.

    When one key occurs several times (a self-join, a CTE referenced twice,
    a scan repeated under an Append), the occurrence with the LARGEST actual
    is kept: it is the one whose misestimate can actually cost time, and it
    is the one both planners will agree is the interesting instance.
    """
    out = {}
    for root in roots:
        annotate_relsets(root)
        for n in walk(root):
            if n.pseudo or not n.relset:
                continue
            if want_actual and n.arows is None:
                continue                      # `never executed`
            key = tuple(sorted(n.relset))
            rec = {'est': n.erows, 'actual': n.arows, 'node': n.name.split('  ')[0].strip()}
            prev = out.get(key)
            if prev is None:
                out[key] = rec
            elif want_actual and (rec['actual'] or 0) > (prev['actual'] or 0):
                out[key] = rec
            elif not want_actual and rec['est'] > prev['est']:
                out[key] = rec
    return out


def load_pg(pgdir):
    """PG 18.3's estimates per query, keyed identically.

    The PG captures carry no ANALYZE, so only `est` is read from them. Their
    actuals come from goopg's run — same data, same queries, so the actual
    row count of a given relation set is a property of the workload, not of
    the planner that computed it.
    """
    pg = {}
    for fn in os.listdir(pgdir):
        m = re.fullmatch(r'Q(\d+)\.txt', fn)
        if not m:
            continue
        with open(os.path.join(pgdir, fn), errors='replace') as fh:
            pg[int(m.group(1))] = collect(parse(fh.read()), want_actual=False)
    return pg


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('capture')
    ap.add_argument('pgdir')
    ap.add_argument('--json')
    ap.add_argument('--baseline')
    ap.add_argument('--write-baseline')
    # TOL: how much worse than PostgreSQL goopg may be on a node before the
    # node is a finding. 2.0 is deliberately generous — the defects this gate
    # is aimed at are three to four ORDERS OF MAGNITUDE, and a tight
    # tolerance would turn the gate into the change-detector that
    # `make plan-gate MODE=semantic-cost` already is.
    ap.add_argument('--tol', type=float, default=2.0)
    # FLOOR: absolute q-error below which nothing is reported however much
    # worse than PG it is. A node PG estimates exactly (q=1) and goopg
    # estimates at q=3 is not a five-orders-of-magnitude defect.
    ap.add_argument('--floor', type=float, default=10.0)
    ap.add_argument('--top', type=int, default=25)
    args = ap.parse_args()

    with open(args.capture, errors='replace') as fh:
        text = fh.read()
    blocks = re.split(r'^===== Q(\d+) =====$', text, flags=re.M)
    pg = load_pg(args.pgdir)

    findings, scored, unmatched = [], 0, 0
    queries = []
    for i in range(1, len(blocks), 2):
        q = int(blocks[i])
        queries.append(q)
        goopg = collect(parse(blocks[i + 1]), want_actual=True)
        pgq = pg.get(q, {})
        for key, rec in goopg.items():
            scored += 1
            gq = qerror(rec['est'], rec['actual'])
            pgrec = pgq.get(key)
            if pgrec is None:
                unmatched += 1
                pq = None
            else:
                pq = qerror(pgrec['est'], rec['actual'])
            # An unmatched key is scored against an absolute bar only. PG
            # having no node with this relation set usually means the two
            # planners chose different join orders, which is a plan-shape
            # question, not an estimate one — so the bar there is loose.
            bar = args.floor if pq is None else max(args.floor, pq * args.tol)
            if gq > bar:
                findings.append({
                    'q': q, 'key': list(key), 'node': rec['node'],
                    'est': rec['est'], 'actual': rec['actual'],
                    'qerr': round(gq, 1),
                    'pg_est': pgrec['est'] if pgrec else None,
                    'pg_qerr': round(pq, 1) if pq is not None else None,
                })

    findings.sort(key=lambda f: -f['qerr'])
    ids = sorted('Q%d:%s' % (f['q'], '+'.join(f['key'])) for f in findings)

    print('capture:   %s' % args.capture)
    print('pg plans:  %s' % args.pgdir)
    print('queries:   %d   nodes scored: %d   unmatched-in-PG: %d'
          % (len(queries), scored, unmatched))
    print('bar:       qerr > max(%.1f, PG_qerr * %.1f)' % (args.floor, args.tol))
    print('FINDINGS:  %d' % len(findings))
    print()
    hdr = ('%5s %10s %12s %10s %10s %10s  %-22s %s'
           % ('Q', 'qerr', 'goopg est', 'ACTUAL', 'pg est', 'pg qerr', 'node', 'relset'))
    print(hdr)
    print('-' * len(hdr))
    for f in findings[:args.top]:
        print('%5s %10.1f %12d %10.0f %10s %10s  %-22s %s'
              % ('Q%d' % f['q'], f['qerr'], f['est'], f['actual'],
                 f['pg_est'] if f['pg_est'] is not None else '-',
                 ('%.1f' % f['pg_qerr']) if f['pg_qerr'] is not None else '-',
                 f['node'][:22], '+'.join(f['key'])[:60]))
    if len(findings) > args.top:
        print('  ... %d more (see --json)' % (len(findings) - args.top))

    if args.json:
        with open(args.json, 'w') as fh:
            json.dump({'queries': queries, 'scored': scored,
                       'tol': args.tol, 'floor': args.floor,
                       'findings': findings}, fh, indent=1)

    if args.write_baseline:
        with open(args.write_baseline, 'w') as fh:
            fh.write('# EA parity ratchet baseline — one line per finding.\n')
            fh.write('# tol=%.1f floor=%.1f\n' % (args.tol, args.floor))
            for i in ids:
                fh.write(i + '\n')
        print('\nbaseline written: %s (%d entries)' % (args.write_baseline, len(ids)))
        return 0

    if args.baseline:
        with open(args.baseline) as fh:
            base = set(l.strip() for l in fh
                       if l.strip() and not l.startswith('#'))
        cur = set(ids)
        new = sorted(cur - base)
        fixed = sorted(base - cur)
        print()
        print('RATCHET vs %s' % args.baseline)
        print('  baseline findings: %d   current: %d' % (len(base), len(cur)))
        for f in fixed:
            print('  FIXED    %s' % f)
        for f in new:
            print('  NEW      %s' % f)
        if new:
            print('\nEA-RATCHET: FAIL (%d new estimate finding(s))' % len(new))
            return 1
        print('\nEA-RATCHET: PASS%s'
              % (' (%d fixed — re-pin the baseline)' % len(fixed) if fixed else ''))
        return 0

    return 0


if __name__ == '__main__':
    sys.exit(main())
